enhancements/cluster-logging/forward-to-loki.md

```diff
@@ -0,0 +1,221 @@
+- Describe how expected query patterns correspond to Loki labels and JSON payload
+
+### Non-Goals
+
+The following may be covered in future proposals, but are out of scope here:
+
+- Query tools.
+- Optimizing/reducing the JSON payload format.
+- Alternate payload formats.
+- Use of collectors other than `fluentd` (e.g. promtail, fluentbit)
+
+## Proposal
+
+### Summary of Loki labels
+
+Loki and Elasticsearch take different approaches to indexing and search.
+Elasticsearch does content parsing and indexing _during ingest_ to optimize the _query_ path.
+Loki _optimizes ingest_ by using simpl key-value `labels` during ingest.
+Content parsing happens at _query_ time, to move work out of the ingest path.
+
+The benefit is faster ingest, which reduces the risk of dropping logs.
+The trade-off is that need to carefully choose what to use as labels.
+
+For more details:
+- [Guide to labels in loki](https://grafana.com/blog/2020/08/27/the-concise-guide-to-labels-in-loki/)
+- [Out-of-order streams issue]([https://github.com/grafana/loki/issues/1544)
+
+To summarize the key points:
+- Each unique combination of labels defines an ordered Loki _stream_.
+  - Too many streams degrades performance.
+  - _Cardinality_ refers to the number of unique label combinations.
+- The cardinality of `kubernetes.pod_name` was found to be
+  - too high in clusters with _very_ rapid turn-over of short-lived pods.
+  - acceptable in other (possibly more typical) clusters with high log data rates.
+- Minimize the _total number_ of labels per stream
+  - There is a limit of 30 max, fewer labels == better performance.
+- Log streams must be _ordered by collection timestamp_.
+  - Must not allow distinct nodes with separate clocks to contribute to a single stream.
+
+All logging meta-data will still be included in log records.
+Labels do not need to _identify the source of logs_, only to _partition the search space_.
+At query time labels reduce the search space, then Loki uses log content to complete the search.
+
+**Note**: Loki [label names][] must match the regex `[a-zA-Z_:][a-zA-Z0-9_:]*`.
+Meta-data names are translated to Loki labels by replacing illegal characters with '_'.
+
+Nested JSON objects are referenced by converting their JSON path to a label names.
+For example, if `kubernetes.namespace_name` is a label, then this JSON log record:
+
+``` json
+{"kubernetes.namespace_name":"foo","kubernetes":{"labels":{"foo":"bar"}}
+```
+would match this LogQL filter: 
+
+``` logql
+{ kubernetes_namespace_name="foo" } | json | kubernetes_labels_foo == "bar"
+```
+### Default Loki Labels
+
+The proposed default label set is:
+
+- `log_type`: one of "application", "infrastructure", "audit"
+- `cluster`: for muti-cluster deployments (see Open Questions).
+- `kubernetes.host`: aka node, ensures ordered streams.
+- `kubernetes.namespace_name`: likely to be most common known factor in a search.
+- `kubernetes.container_name`: low cardinality, usually well-known
```

Comment:
@alanconway This label is strange to me ... I understand that this is low cardinality ... but on the flip side if there is no indexing to pod name ... why the container name is intersting .... in any event the code need to scan to find all pod names ... so it doesn't sound expensive to me to look for container_name

Reply:
@alanconway there are two labels that I was expecting and I do not see in the default list::

(1) `app` which is very common and might allow fast focus on specific application ... this is similar in sense to namespace because this is "logical" label but the granularity of application (i.e. cardinality) is higher. On the other hand the cardinality is limited to the applications in the cluster so no reason to expect huge number of those
 
(2) `tier` same as (1) with a bit higher cardinality maybe a bit similar to `container_name` but in the logical space and not in the physical space

Reply:
> @alanconway This label is strange to me ... I understand that this is low cardinality ... but on the flip side if there is no indexing to pod name ... why the container name is intersting .... in any event the code need to scan to find all pod names ... so it doesn't sound expensive to me to look for container_name

Good question: I'm uncertain about it. The idea is that container name is a constant for pods created by deployments, even when the pod names are randomized. So it might be useful to locate pods that are part of a multi-namespace application. I think it needs a bit more justification though based on looking at real clusters.

Reply:
> @alanconway there are two labels that I was expecting and I do not see in the default list::
> 
> (1) `app` which is very common and might allow fast focus on specific application ... this is similar in sense to namespace because this is "logical" label but the granularity of application (i.e. cardinality) is higher. On the other hand the cardinality is limited to the applications in the cluster so no reason to expect huge number of those

Again you spot my every uncertainty! First draft of this did include kubernetes.labels.app and kubernetes.labels.deploymentconfig in the defaults. However they can be enabled using `labelKeys` so I took them out.
I can probably be argued into putting "app" back  into defaults, as it is a crucial one for cross-namespace apps.

> (2) `tier` same as (1) with a bit higher cardinality maybe a bit similar to `container_name` but in the logical space and not in the physical space

What's "tier"?

Reply:
@alanconway my mistake ... I should have used the correct label which is `app.kubernetes.io/component` but actually I assume your answer to that and thoughts will be similar that this can be part of default but maybe better to get more experiance and for now can be added as part of the configuration

Reply:
@alanconway What do you think of putting the pod name in `labelKeys` per default, i.e. `labelKeys = []string{"kubernetes_pod_name"}`. Users will basically have it in default installation and have the choice to remove it later, if they have more flaky workloads, disruptive environment or more batch-like workloads (like DPTP).

Reply:
@periklis that's the other thing I'm  uncertain of. Loki guys also suggested this. Main considerations I see:

1. We know it was a problem for DPTP, we _believe_ DPTP is unusual. Loki folks claim to have included it without problems but we don't have details of those clusters.
2. When searching for a named pod, you clearly know the namespace, so you're only searching per namespace.
3. For post-mortem bug-hunts, a responsive search for specific pods is important.
4. For routine cluster monitoring, specific pod search is not important at all,  you want big cluster-wide searches for specific log types, severities or contents.

So: do we let people opt in to improve performance, or tell them to opt out after a disaster? I think you have the best grasp of the trade-off, but I'd love to have numbers. In particular I guess that including pod name is good for 3. but is it bad for 4.? What are your thoughts?

Reply:
I would prefer that people opt out. The problem DPTP had is something we'll address over time.

Reply:
@alanconway I am pro "opt-out" without having numbers to prove this. The pod name sets from a user experience a high expectation. The user could achieve the same with the cluster node and the container name assuming bin packing is biting (having 2 replicas of the container on the same node). 

@cyriltovena Would you mind to share some insights from your experience if possible, e.g.
- The amount of pod name labels used in various use cases (.g. 0.5TB/day, 1TB/day, 2TB/day ingestion)
- The impact of this pod name label volumes on botldb index size?

Reply:
I have a cluster that supports 13k pod names (consistent) for 10TB per day. The stream churns for that cluster over a day is 2millions. (stream added + stream removed).

Again we are working at improving everyday on that matter.

Reply:
It has been updated to be opt-out.